pkg/engine/runtime/kubernetes/kubernetes_runtime.go

```diff
@@ -543,68 +574,51 @@ func convertString2Unstructured(yamlContent []byte) (*unstructured.Unstructured,
 	return obj, gvk, nil
 }
 
-// WatchBySelector watch resources by gvk and filter by selector
-func (k *KubernetesRuntime) WatchBySelector(
-	ctx context.Context,
-	o *unstructured.Unstructured,
-	gvk schema.GroupVersionKind,
-	id, labelStr string,
-) (<-chan k8swatch.Event, error) {
-	clientForResource, err := buildDynamicResource(k.client, k.mapper, &gvk, id, o.GetNamespace())
-	if err != nil {
-		return nil, err
-	}
-
-	w, err := clientForResource.Watch(ctx, metav1.ListOptions{LabelSelector: labelStr})
-	if err != nil {
-		return nil, err
-	}
-
-	return doWatch(ctx, w, nil), nil
-}
-
 // WatchByRelation watched resources by giving gvk if related() return true
 func (k *KubernetesRuntime) WatchByRelation(
 	ctx context.Context,
 	cur *unstructured.Unstructured,
 	gvk schema.GroupVersionKind,
+	amount int,
 	related func(watched, cur *unstructured.Unstructured) bool,
-) (<-chan k8swatch.Event, *unstructured.Unstructured, error) {
+) ([]KubernetesWatchEvent, error) {
 	mapping, err := k.mapper.RESTMapping(gvk.GroupKind(), gvk.Version)
 	if err != nil {
-		return nil, nil, err
+		return nil, err
 	}
 
 	clientForResource := k.client.Resource(mapping.Resource).Namespace(cur.GetNamespace())
 	w, err := clientForResource.Watch(context.TODO(), metav1.ListOptions{})
 	if err != nil {
-		return nil, nil, err
+		return nil, err
 	}
-
-	var next *unstructured.Unstructured
-
-	eventCh := doWatch(ctx, w, func(watched *unstructured.Unstructured) bool {
+	// Get all the watch events of the dependents of a resource
+	events := doWatch(ctx, w, amount, func(watched *unstructured.Unstructured) bool {
 		ok := related(watched, cur)
-		if ok {
-			next = watched
-		}
 		return ok
 	})
-	if eventCh == nil {
+	if len(events) == 0 {
 		err = fmt.Errorf("failed to get the event channel for watching related resources of %s with kind of %s",
 			cur.GetName(), cur.GetKind())
 	}
 
-	return eventCh, next, err
+	return events, err
 }
 
-// doWatch send watched object if check ok
-func doWatch(ctx context.Context, watcher k8swatch.Interface, checker func(watched *unstructured.Unstructured) bool) <-chan k8swatch.Event {
-	// Buffered channel, store new event
-	resultCh := make(chan k8swatch.Event, 1)
-	// Wait for the first watched obj
-	first := true
+// doWatch send KubernetesWatchEvent Slice which contains all the watch channels and their belonging resources if check ok
+func doWatch(ctx context.Context, watcher k8swatch.Interface, amount int, checker func(watched *unstructured.Unstructured) bool) []KubernetesWatchEvent {
+	events := []KubernetesWatchEvent{}
+	eventMap := make(map[string]chan k8swatch.Event)
 	signal := make(chan struct{})
+	timeoutCh := make(chan struct{})
+	if amount == -1 {
+		go func() {
```

Comment:
Kusion may miss the K8s event if the api-server processes longer than 2 seconds, I think it might be better to wait for the timeout after receiving the first event.

Reply:
> Kusion may miss the K8s event if the api-server processes longer than 2 seconds, I think it might be better to wait for the timeout after receiving the first event.

Will fix.